java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java

```diff
@@ -0,0 +1,256 @@
+          argumentValues, "Parameter {0} to this constructor is marked \"{1}\" but null is passed.");
+      }
+    }
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree syntaxTree) {
+      Symbol symbol = syntaxTree.symbol();
+      if (symbol.isMethodSymbol()) {
+        Arguments arguments = syntaxTree.arguments();
+        List<SymbolicValue> argumentValues = new ArrayList<>(programState.peekValues(arguments.size() + 1));
+        argumentValues.remove(0);
+        // Arguments of method invocation are in reverse order on the stack, unlike constructors
+        Collections.reverse(argumentValues);
+        JavaSymbol.MethodJavaSymbol methodSymbol = (JavaSymbol.MethodJavaSymbol) symbol;
+        checkNullArguments(syntaxTree, methodSymbol.getParameters(),
+          argumentValues, "Parameter {0} to this call is marked \"{1}\" but null is passed.");
+      }
+    }
+
+    protected void checkNullArguments(Tree syntaxTree, Scope parameters, List<SymbolicValue> argumentValues, String message) {
+      if (parameters != null) {
+        List<JavaSymbol> scopeSymbols = parameters.scopeSymbols();
+        int parametersToTest = argumentValues.size();
+        if (scopeSymbols.size() < parametersToTest) {
+          // The last parameter is a variable length argument: the non-null condition does not apply to its values
+          parametersToTest = scopeSymbols.size() - 1;
+        }
+        for (int i = 0; i < parametersToTest; i++) {
+          checkNullArgument(syntaxTree, message, scopeSymbols, argumentValues, i);
+        }
+      }
+    }
+
+    protected void checkNullArgument(Tree syntaxTree, String message, List<JavaSymbol> scopeSymbols, List<SymbolicValue> argumentValues, int i) {
+      String nonNullAnnotation = nonNullAnnotation(scopeSymbols.get(i));
+      if (nonNullAnnotation != null) {
+        Constraint constraint = programState.getConstraint(argumentValues.get(i));
+        if (constraint != null && constraint.isNull()) {
+          reportIssue(syntaxTree, message, Integer.valueOf(i + 1), nonNullAnnotation);
+        }
+      }
+    }
+  }
+
+  private class PostStatementVisitor extends AbstractStatementVisitor {
+
+    protected PostStatementVisitor(CheckerContext context) {
+      super(context);
+    }
+
+    @Override
+    public void visitReturnStatement(ReturnStatementTree tree) {
+      Tree parent = tree.parent();
+      while (!parent.is(Tree.Kind.METHOD)) {
+        parent = parent.parent();
+        if (parent == null) {
+          // This occurs when the return statement is within a constructor
+          return;
+        }
+      }
+      MethodTree mTree = (MethodTree) parent;
+      String nonNullAnnotation = nonNullAnnotation(mTree.symbol());
+      if (nonNullAnnotation != null) {
+        if (isLocalExpression(tree.expression())) {
+          checkReturnedValue(tree, nonNullAnnotation);
+        }
+      }
+    }
+
+    private boolean isLocalExpression(ExpressionTree expression) {
```

Comment:
Looks very close from the method `isNonLocalStorage` that you implemented in 'UnclosedResourcesCheck', except that this time you are checking if it's a local variable. The two methods should be renamed to match this behavior.